misaligned.c

```diff
@@ -1,21 +1,85 @@
 #include <stdio.h>
 #include <assert.h>
+#include <string.h> 
+
+const char* majorColor[] = {"White", "Red", "Black", "Yellow", "Violet"};
+const char* minorColor[] = {"Blue", "Orange", "Green", "Brown", "Slate"};
+
+// Structure to return the color pair information
+typedef struct{
+    int colorCode;
+    const char *majorColor;
+    const char *minorColor;
+}colorPair;
+
+colorPair getColorCode(int i, int j){
+    colorPair colorPair;
+    // Deriving colorPair.colorCode, colorPair.majorColor and colorPair.minorColor are
+    // Done as per the base code. No modification in logic was done
+    colorPair.colorCode = i * 5 + j;
+    colorPair.majorColor = majorColor[i];
+    colorPair.minorColor = minorColor[i];
+    return colorPair;
+}
 
 int printColorMap() {
-    const char* majorColor[] = {"White", "Red", "Black", "Yellow", "Violet"};
-    const char* minorColor[] = {"Blue", "Orange", "Green", "Brown", "Slate"};
     int i = 0, j = 0;
+    colorPair colorPair;
     for(i = 0; i < 5; i++) {
         for(j = 0; j < 5; j++) {
-            printf("%d | %s | %s\n", i * 5 + j, majorColor[i], minorColor[i]);
+            colorPair = getColorCode(i, j);
+            printf("%d | %s | %s\n", colorPair.colorCode, colorPair.majorColor, colorPair.minorColor);
         }
     }
     return i * j;
 }
 
-int main() {
+int main(){
+    // Test case 1: Validate the number of color pairs printed
     int result = printColorMap();
     assert(result == 25);

```

Comment:
Testing color codes is not adequate.
Number of combinations is necessary but not sufficient.
The tests should include how they are printed and their alignment also. Please consider the pair with single and double digits and its effect on alignment.

Reply:
Agreed. Alignment also shall be tested along with color pair combinations

Reply:
I have no idea how I can test the alignment since the code says the format is like "%d | %s | %s\n". Should I assume the same format for each line should there be a tab between each field to make nicer format? What should be the requirement or assumption here?

I assumed, color pair value itself is more important than alignment!!!

Reply:
Color pair is important, but let's understand the requirements. We need aligned output.
Just for understanding let's say output is as follows,
1 | white | orange
20 | white | blue
The pipe symbol is not aligned. You need to consider single digit and double digit pair number and variable length color name. 

You may assume certain things and accordingly assert to ascertain the outcome.

Important thing here is, you should be able to define the requirements as test cases and asserts without writing the implementation yet 🙂

Let me know if you have any doubts 👍